### PR TITLE
Support mutation of component attributes via component metadata rules

### PR DESCRIPTION
**Pull request by melix**
### Context

This implements #3159.

### CODE REVIEW

**Comment by melix** [comment]
So component level attribute mutation is supported in `32598fb`. It raises some questions:

- first, we need to pass the attribute factory down to `VariantBackedConfigurationMetadata`, so that when selection occurs, we can properly choose between variants that will carry both the _component_ attributes and _variant_ attributes. In other words, when we get the `VariantBackedConfigurationMetadata` attributes, we see a merged view of the processed component attributes and processed variant attributes
- should we see the `status` as an attribute possibly mutated? Knowing that we already have `setStatus` in rules, we could consider deprecating `setStatus` in favor of mutation, or, for practical reasons, delegate the first one to the attributes version (because one doesn't want to know the internal name of the status attribute)
- should selection be consistent independently of where the attribute is defined? For example, it doesn't actually matter if you mutate the `component` attribute, or if you do this for all variants. One reason to do it on the component level is that you don't care about variants there, it's applied unconditionally (and you don't have to know the variant names)
- while mutation is supported, its use is actually limited since we are still unable to select appropriately with dynamic version selectors: attributes are verified _after_ selection, meaning that the selected version in a range will be eventually rejected because it doesn't match the attributes.

**Comment by melix** [comment]
This is now ready for review.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/VariantAttributesRulesIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+ * limitations under the License.
+ */
+package org.gradle.integtests.resolve
+
+import org.gradle.integtests.fixtures.GradleMetadataResolveRunner
+import org.gradle.integtests.fixtures.RequiredFeature
+import org.gradle.integtests.fixtures.RequiredFeatures
+import spock.lang.Unroll
+
+class VariantAttributesRulesIntegrationTest extends AbstractModuleDependencyResolveTest {
```
Comment (bigdaz) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/VariantAttributesRulesIntegrationTest.groovy]: :o: These tests will benefit from adding the variant details to the `ResolutionResult` API. We should remember to update the tests when this happens: add a story to remind us?

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentMetadataDetails.java [hunk 2]
```diff
@@ -63,4 +64,13 @@
...
     void withVariant(String name, Action<VariantMetadata> action);
+
+    /**
+     * Adjust the component-level attributes.
+     *
+     * @param action the adjustment action to be executed on attributes
+     *
+     * @since 4.5
+     */
+    void withAttributes(Action<? super AttributeContainer> action);
```
Comment (bigdaz) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentMetadataDetails.java]: :x: Should also be able to read attributes via `ComponentMetadata` supertype. Then the attributes can be consumed in `ComponentSelectionRules`

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentAttributesRulesIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            failure.assertThatCause(containsNormalizedString("Found org.gradle.usage 'unknownRuntimeVariant' but wasn't required"))
+        }
+
+        where:
+        componentLevel << [true, false]
+    }
+
+    // This is eventually what we want, but version listing is not wired properly for this
+    @Ignore
+    def "can use a component metadata rule to infer quality attribute"() {
```
Comment (bigdaz) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentAttributesRulesIntegrationTest.groovy]: :x: Should add test coverage for consuming a component attribute via `ComponentSelectionRules`.

**Comment by bigdaz** [comment]
> first, we need to pass the attribute factory down to VariantBackedConfigurationMetadata, so that when selection occurs, we can properly choose between variants that will carry both the component attributes and variant attributes. In other words, when we get the VariantBackedConfigurationMetadata attributes, we see a merged view of the processed component attributes and processed variant attributes

As I mentioned in the review, I think the attribute factory service should be provided via injection, and should not pollute the API.

> should we see the status as an attribute possibly mutated? Knowing that we already have setStatus in rules, we could consider deprecating setStatus in favor of mutation, or, for practical reasons, delegate the first one to the attributes version (because one doesn't want to know the internal name of the status attribute)

Yes, the `getStatus()` and `setStatus()` methods should be backed by attributes, I think.

> should selection be consistent independently of where the attribute is defined? For example, it doesn't actually matter if you mutate the component attribute, or if you do this for all variants. One reason to do it on the component level is that you don't care about variants there, it's applied unconditionally (and you don't have to know the variant names)

I think it's OK for now that variant attributes are not considered in component selection. Having the same attribute on every variant is not (yet) the same as having the attribute on the component.

> while mutation is supported, its use is actually limited since we are still unable to select appropriately with dynamic version selectors: attributes are verified after selection, meaning that the selected version in a range will be eventually rejected because it doesn't match the attributes.

A component selection rule should have access to the component attributes, and should be able to make decisions based on these. So attribute mutation at the component level makes sense in this case.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentAttributesRulesIntegrationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+import spock.lang.Ignore
+import spock.lang.Unroll
+
+@RequiredFeatures(
+    @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "true")
+)
+class ComponentAttributesRulesIntegrationTest extends AbstractModuleDependencyResolveTest {
+
+    @Unroll
+    def "succeeds if attribute is added to component via component metadata rules (fix = #fixApplied)"() {
```
[the remark this answers is not in the document]
Comment (bigdaz) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentAttributesRulesIntegrationTest.groovy]: Need test coverage for the interaction of `ComponentMetadata.status` and the status attribute. I think the `status` property should be backed by the attribute (or vice-versa).

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentMetadataDetails.java (hunk 2 above)
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentMetadataDetails.java]: Right, I wasn't sure if we wanted to expose such an API there. I will add it.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/DefaultArtifactSet.java [hunk 5]
```diff
@@ -99,6 +102,7 @@ private static ResolvedVariant toResolvedVariant(VariantMetadata variant, Module
             }
             resolvedArtifacts.add(resolvedArtifact);
         }
+        attributes = componentMetadataRules.applyVariantAttributeRules(attributes);
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/DefaultArtifactSet.java]: That's what I meant when I said we had to apply the rules twice. I agree there's something fishy, and I also would like to understand why.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ComponentMetadataRules.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.api.internal.attributes.ImmutableAttributesFactory;
+import org.gradle.internal.component.model.DependencyMetadataRules;
+import org.gradle.internal.component.model.VariantAttributesRules;
+import org.gradle.internal.reflect.Instantiator;
+import org.gradle.internal.typeconversion.NotationParser;
+
+import java.util.List;
+
+public class ComponentMetadataRules {
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ComponentMetadataRules.java]: Renamed to `VariantMetadataRules`, I wanted to do this actually and just forgot.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/ComponentResolveMetadata.java [hunk 7]
```diff
@@ -90,4 +94,6 @@
     String getStatus();
 
     List<String> getStatusScheme();
+
+    Map<String, ComponentMetadataRules> getComponentMetadataRules();
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/ComponentResolveMetadata.java]: It's used for artifact resolution. It's probably the same underlying reason as to why we have to process attributes during artifact resolution.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/VariantMetadata.java [hunk 8]
```diff
@@ -25,6 +25,8 @@
  * Metadata for a basic variant of a component, that defines only artifacts and no dependencies.
  */
 public interface VariantMetadata {
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/VariantMetadata.java]: I didn't want to add more to this PR, but we agreed with @jjohannes to perform renames as a separate PR. We don't want to block too much work on this.

[the remark this answers is not in the document]
Comment (jjohannes) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/VariantMetadata.java]: There are also at least one more internal interfaces which should be renamed: `DependenyMetadata`.

I added this to: #3732
There are some other structure/naming concerns with the public interfaces as well.

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentMetadataDetails.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentMetadataDetails.java]: fixed

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/VariantMetadata.java [hunk 9]
```diff
@@ -42,4 +43,14 @@
...
+
+    /**
+     * Registers a rule that will execute on the variant attributes. This allows adding
+     * or modifying variant attributes, but not removing them.
+     *
+     * @param action the action that performs the attributes adjustments
+     *
+     * @since 4.5
+     */
+    void withAttributes(Action<? super AttributeContainer> action);
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/VariantMetadata.java]: fixed

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ModuleComponentResolveMetadata.java [hunk 10]
```diff
@@ -54,4 +55,6 @@
      * Returns the hash of the resource(s) from which this metadata was created.
      */
     HashValue getContentHash();
+
+    AttributeContainer getAttributes();
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ModuleComponentResolveMetadata.java]: I looked at the difference between `ModuleComponentResolveMetadata` and `ComponentResolveMetadata`, and it's actually unclear to me why we need both. Can't they be merged together?

**Pull request merged**